pkg/apis/authorization/types.go

```diff
@@ -0,0 +1,120 @@
+// to check whether they can perform an action
+type SelfSubjectAccessReview struct {
+	unversioned.TypeMeta
+
+	// Spec holds information about the request being evaluated.
+	Spec SelfSubjectAccessReviewSpec
+
+	// Status is filled in by the server and indicates whether the request is allowed or not
+	Status SubjectAccessReviewStatus
+}
+
+// LocalSubjectAccessReview checks whether or not a user or group can perform an action in a given namespace.
+// Having a namespace scoped resource makes it much easier to grant namespace scoped policy that includes permissions
+// checking.
+type LocalSubjectAccessReview struct {
+	unversioned.TypeMeta
+
+	// Spec holds information about the request being evaluated.  spec.namespace must be equal to the namespace
+	// you made the request against.  If empty, it is defaulted.
+	Spec SubjectAccessReviewSpec
+
+	// Status is filled in by the server and indicates whether the request is allowed or not
+	Status SubjectAccessReviewStatus
+}
+
+type ResourceAttributes struct {
+	// Namespace is the namespace of the action being requested.  Currently, there is no distinction between no namespace and all namespaces
+	// "" (empty) is defaulted for LocalSubjectAccessReviews
+	// "" (empty) is empty for cluster-scoped resources
+	// "" (empty) means "all" for namespace scoped resources from a SubjectAccessReview or SelfSubjectAccessReview
+	Namespace string
+	// Verb is a kubernetes resource API verb, like: get, list, watch, create, update, delete, proxy.  "*" means all.
+	Verb string
+	// Group is the API Group of the Resource.  "*" means all.
+	Group string
+	// Version is the API Version of the Resource.  "*" means all.
+	Version string
+	// Resource is one of the existing resource types.  "*" means all.
+	Resource string
+	// Subresource is one of the existing resource types
+	Subresource string
+	// Name is the name of the resource being requested for a "get" or deleted for a "delete". "" (empty) means all.
+	Name string
+}
+
+type NonResourceAttributes struct {
+	// Path is the URL path of the request
+	Path string
+	// Verb is the standard HTTP verb
+	Verb string
+}
+
+// SubjectAccessReviewSpec is a description of the access request.  Exactly one of ResourceAttributes
+// and NonResourceAttributes must be set
+type SubjectAccessReviewSpec struct {
+	// ResourceAttributes describes information for a resource access request
+	ResourceAttributes *ResourceAttributes
+	// NonResourceAttributes describes information for a non-resource access request
+	NonResourceAttributes *NonResourceAttributes
+
+	// User is the user you're testing for.  If neither user, nor groups is set, then the check is performed
+	// for the current user and group.
+	User string
+	// Groups is the groups you're testing for.  If neither user, nor groups is set, then the check is performed
+	// for the current user and group.
+	Groups []string
```

Comment:
I vaguely remember discussing this...

pros of groupifying server-side:
* allows things outside the auth-layer to check actual access for a user without needing to know the user's group membership

cons of groupifying server-side:
* tightly couples the SAR implementation to the auth layer
* as you point out, makes it more difficult to support auth methods that communicate group membership at login time (like SAML or auth proxies with extended attributes)

other thoughts:
* what if the user cannot be groupified? for act-as, we had said we thought that should be a failure, but that doesn't seem right here
* I would want the ability to check just a user's access (e.g. `{user:"bob",groups:[]}`), but leaving groups `nil` to request groupification from the server seems super-subtle